grin/src/Transformations/ExtendedSyntax/Optimising/CaseCopyPropagation.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE LambdaCase #-}
+module Transformations.ExtendedSyntax.Optimising.CaseCopyPropagation where
+
+import Data.Map (Map)
+import Data.Functor.Foldable
+
+import qualified Data.Map as Map
+
+import Control.Monad.State
+
+import Grin.ExtendedSyntax.Grin
+import Transformations.ExtendedSyntax.Names
+import Transformations.ExtendedSyntax.Util (cataM)
+
+
+-- NOTE: ~ Maybe Tag
```

Comment:
I prefer to encode with descriptive types and trying to avoid boolean blindness.